server/auth/providers/oauth2.js

```diff
@@ -0,0 +1,123 @@
+// @flow
+import passport from "@outlinewiki/koa-passport";
+import fetch from "fetch-with-proxy";
+import Router from "koa-router";
+import { Strategy } from "passport-oauth2";
+import accountProvisioner from "../../commands/accountProvisioner";
+import env from "../../env";
+import {
+  OAuth2MalformedUserInfoError,
+  AuthenticationError,
+} from "../../errors";
+import passportMiddleware from "../../middlewares/passport";
+import { getAllowedDomains } from "../../utils/authentication";
+import { StateStore } from "../../utils/passport";
+
+const router = new Router();
+const providerName = "oauth2";
+const OIDC_DISPLAY_NAME = process.env.OIDC_DISPLAY_NAME || "OAuth2";
+const OIDC_CLIENT_ID = process.env.OIDC_CLIENT_ID;
+const OIDC_CLIENT_SECRET = process.env.OIDC_CLIENT_SECRET;
+const OIDC_AUTH_URI = process.env.OIDC_AUTH_URI;
+const OIDC_TOKEN_URI = process.env.OIDC_TOKEN_URI;
+const OIDC_USERINFO_URI = process.env.OIDC_USERINFO_URI;
+const OIDC_SCOPES = process.env.OIDC_SCOPES || "";
+const allowedDomains = getAllowedDomains();
+
+export const config = {
+  name: OIDC_DISPLAY_NAME,
+  enabled: !!OIDC_CLIENT_ID,
+};
+
+const scopes = OIDC_SCOPES.split(" ");
+
+Strategy.prototype.userProfile = async function (accessToken, done) {
+  try {
+    const response = await fetch(OIDC_USERINFO_URI, {
+      credentials: "same-origin",
+      headers: {
+        Authorization: `Bearer ${accessToken}`,
+      },
+    });
+
+    try {
+      return done(null, await response.json());
+    } catch (ex) {
+      return done(new Error("Failed to parse user profile"));
+    }
+  } catch (err) {
+    return done(new Error("Failed to fetch user profile"));
```

Comment:
Should probably log the error here at the very least otherwise debugging this will be very difficult

Reply:
Solved this by passing the existing errors forward.